### ISSUE TITLE
Give fcase an "as factor" option

### ISSUE THREAD

**Issue by markdanese** [Rdatatable/data.table#5327]
I looked for this as an open issue and did not see it.  I use fcase to create categorical variables.  These are almost always character.  However, most of the time, I need them as factors, which requires me to retype the same factor levels, often in the same order I used them in fcase.  My request is to add an option to have fcase create the factor variable directly, using the order from fcase itself.  See below for a rough example.  

If you wanted to be fancy, you could allow for other options too, like factor = "reverse" (to set the factor order level in the opposite sequence from the fcase statement) or even allow for a custom order by using numbers (e.g., factor = c(2, 4, 1, 3)).  But just the basic version would be very helpful.

dt[, 
      newvar :=
          fcase(
              logic1 = "category 1",
              logic2 = "category 2",
              logic3 = "category 3",
              default = "category 4"
              )]
dt[, newvar := factor(newvar, levels = c("category 1", "category 2", "category 3", "category 4")]

would become
dt[, 
      newvar :=
          fcase(
              logic1 = "category 1",
              logic2 = "category 2",
              logic3 = "category 3",
              default = "category 4",
              factor = TRUE
              )]

I know you all have tons of more important things on your plate.  Regardless of whether this can be done, thanks for all your help in keeping data.table going forward.

**Comment by jangorecki** [comment]
Order or factor levels, including a reverse order, is nicely addressed by helper function in https://github.com/Rdatatable/data.table/pull/4843
IMO it is better to use this helper rather than adding feature you requested here.

**Comment by jangorecki** [comment]
Using linked PR all needed to address your use case is
fctr(fcase(...)) or fctr(fcase(..., rev=TRUE)).

**Comment by markdanese** [comment]
Thank you.

**Comment by MichaelChirico** [comment]
I'm not sure fctr() solves the issue because the fcase() output can be in any order.

consider

x = c(a, b, c)
fcase(
  x == b, 'first',
  x == a, 'second',
  x == c, 'third'
)

**Comment by jangorecki** [comment]
@MichaelChirico then factor levels will retain order of the input object, which feels more reasonable than sorting levels without being asked for that. @markdanese example does not really stress that.
```r
fctr = function(x, levels=unique(x), ..., rev=FALSE) {
  factor(x, levels=if (rev) rev(levels) else levels, ...)
}
x = c("a", "b", "c")
fcase(
  x == "b", 'first',
  x == "a", 'second',
  x == "c", 'third'
) -> z
fctr(z)
#[1] second first  third 
#Levels: second first third
```

If sorted levels are required then instead of `fctr(z)` one calls `factor(z)`.

**Comment by markdanese** [comment]
I agree that the default should be the actual order in the fcase statement.  A user-selected option to sort could also be useful since sometimes one might construct the fcase statement that optimizes the logic, and not the order.  I suppose that if one wanted to cover all bases, a third option to take a vector of integers for a custom order like c(2, 3, 1) could also be useful.

**Comment by jangorecki** [comment]
My point is to, rather than adding new argument to fcase, **which is only relevant when factor class is returned**, is to ask a user to simply `fcase(...) |> fctr()`.

**Comment by markdanese** [comment]
Makes sense to me.  I misunderstood what fctr() was.
